db/db_impl/db_impl_open.cc

```diff
@@ -1929,4 +1929,305 @@ Status DBImpl::Open(const DBOptions& db_options, const std::string& dbname,
+                                             handles, dbptr);
+}
+
+Status DBImpl::CreateFromDisjointInstances(
+    const MergeInstanceOptions& merge_options, const DBOptions& db_options,
+    const std::string& name,
+    const std::vector<ColumnFamilyDescriptor>& column_families,
+    const std::vector<DB*> instances, std::vector<ColumnFamilyHandle*>* handles,
+    DB** dbptr) {
+  Status s;
+  MergeInstanceOptions mopts(merge_options);
+  // Sanity check.
+  if (column_families.size() == 0 ||
+      column_families.front().name !=
+          ROCKSDB_NAMESPACE::kDefaultColumnFamilyName) {
+    return Status::InvalidArgument("Missing default column family descriptor.");
+  }
+  if (!db_options.create_if_missing) {
+    return Status::InvalidArgument(
+        "create_if_missing must be set to create a DB.");
+  }
+  // Create target DB and column families.
+  s = DB::Open(Options(db_options, column_families.front().options), name,
+               dbptr);
+  if (!s.ok()) {
+    delete *dbptr;
+    *dbptr = nullptr;
+    return s;
+  }
+  WriteBufferManager* write_buffer_manager =
+      db_options.write_buffer_manager.get();
+  DBImpl* new_db_impl = static_cast<DBImpl*>(*dbptr);
+  autovector<uint64_t> cf_ids;
+  autovector<ColumnFamilyData*> new_cfds;
+  if (new_db_impl->total_log_size_ > 0) {
+    s = Status::InvalidArgument("Target DB WAL is not empty.");
+  }
+  for (auto& cf : column_families) {
+    ColumnFamilyHandle* cf_handle;
+    if (s.ok()) {
+      if (handles->size() == 0) {
+        auto* cfd = new_db_impl->versions_->GetColumnFamilySet()->GetDefault();
+        cf_handle =
+            new ColumnFamilyHandleImpl(cfd, new_db_impl, &new_db_impl->mutex_);
+      } else {
+        s = new_db_impl->CreateColumnFamily(cf.options, cf.name, &cf_handle);
+      }
+    }
+    if (s.ok()) {
+      assert(cf_handle);
+      handles->push_back(cf_handle);
+      cf_ids.push_back(cf_handle->GetID());
+      ColumnFamilyData* cfd =
+          new_db_impl->versions_->GetColumnFamilySet()->GetColumnFamily(
+              cf_handle->GetID());
+      new_cfds.push_back(cfd);
+      assert(cfd != nullptr && !cfd->IsDropped());
+
+      VersionStorageInfo& vsi = *cfd->current()->storage_info();
+      if (vsi.num_non_empty_levels() != 0) {
+        s = Status::InvalidArgument("Target DB table data is not empty.");
+      }
+    }
+    if (!s.ok()) {
+      break;
+    }
+  }
+  // Validate and prepare source DBs.
+  autovector<DBImpl*> db_impls;
+  for (auto* db : instances) {
+    DBImpl* db_impl = static_cast<DBImpl*>(db);
+    db_impls.push_back(db_impl);
+    if (db_impl->write_buffer_manager_ != write_buffer_manager) {
+      mopts.merge_memtable = false;
+    }
+    if (s.ok() && db_impl->total_log_size_ != 0) {
+      s = Status::InvalidArgument("DB WAL should be empty.");
+    }
+    if (s.ok() && db_impl->write_buffer_manager_->buffer_size() > 0) {
+      s = Status::InvalidArgument("DB memtable should be empty.");
```

Comment:
The error message seems misleading. buffer_size() > 0 means the memtable has the memory cap.
Also to limit the total memory consumption, we likely will have the limit for memory table.

Reply:
My mistake. I thought it's the used buffer size.

Reply:
Removed. Memtable will just be ignored if `merge_memtable` is false.